File: python-backend/middleware/security.py
"""
middleware/security.py
======================
Security headers middleware for FastAPI/Starlette.

Applies hardened HTTP security headers to every response.
These headers instruct browsers to apply strict security policies.
"""

from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import Response
from config import get_settings


class SecurityHeadersMiddleware(BaseHTTPMiddleware):
    """
    Adds the following security headers to every response:
    - X-Content-Type-Options: Prevents MIME-type sniffing
    - X-Frame-Options: Blocks clickjacking via iframe embedding
    - Referrer-Policy: Limits referrer info sent to other sites
    - X-XSS-Protection: Disables legacy XSS auditor (modern browsers use CSP)
    - Permissions-Policy: Disables camera, mic, geolocation, payment APIs
    - Strict-Transport-Security: Forces HTTPS (production only)
    - X-Download-Options: Prevents IE from executing downloads in site context
    - X-Permitted-Cross-Domain-Policies: Blocks Adobe Flash/Acrobat cross-domain access
    """

    async def dispatch(self, request: Request, call_next) -> Response:
        response = await call_next(request)
        settings = get_settings()

        response.headers["X-Content-Type-Options"] = "nosniff"
        response.headers["X-Frame-Options"] = "DENY"
        response.headers["Referrer-Policy"] = "strict-origin-when-cross-origin"
        response.headers["X-XSS-Protection"] = "0"
        response.headers["X-Download-Options"] = "noopen"
        response.headers["X-Permitted-Cross-Domain-Policies"] = "none"
        response.headers["Permissions-Policy"] = (
            "camera=(), microphone=(), geolocation=(), payment=(), "
            "usb=(), magnetometer=(), gyroscope=(), accelerometer=()"
        )

        # HSTS — only in production (browsers cache this for 1 year)
        if settings.is_production:
            response.headers["Strict-Transport-Security"] = (
                "max-age=31536000; includeSubDomains; preload"
            )

        return response

File: backend/src/middlewares/security.ts
import type { Request, Response, NextFunction } from "express";

// All security headers are set on every response, regardless of route.
// They tell browsers how to handle this content and prevent a range of
// common client-side attacks (XSS, clickjacking, MIME sniffing, etc.).
export function securityHeaders() {
  return (_req: Request, res: Response, next: NextFunction) => {

    // Prevents browsers from MIME-sniffing a response away from the declared Content-Type.
    res.setHeader("X-Content-Type-Options", "nosniff");

    // Stops this site from being embedded in iframes (clickjacking protection).
    res.setHeader("X-Frame-Options", "DENY");

    // Sends the full origin in same-origin requests, only the origin in cross-origin ones.
    res.setHeader("Referrer-Policy", "strict-origin-when-cross-origin");

    // Disables the legacy browser XSS filter — modern CSP is the real defence.
    res.setHeader("X-XSS-Protection", "0");

    // Prevents Adobe Flash / Acrobat from loading cross-domain content.
    res.setHeader("X-Permitted-Cross-Domain-Policies", "none");

    // Stops IE from executing downloaded files in the context of the site.
    res.setHeader("X-Download-Options", "noopen");

    // Locks down hardware access — camera, mic, location, payment are all off.
    res.setHeader(
      "Permissions-Policy",
      "camera=(), microphone=(), geolocation=(), payment=()"
    );

    // Content-Security-Policy: the most important header.
    // This tells browsers exactly which sources are trusted for scripts, styles, images, etc.
    // Adjust these directives if you add new CDNs or external services.
    const csp = [
      // Only load documents from the same origin
      "default-src 'self'",
      // Scripts: same origin + inline scripts (needed by Vite/React)
      "script-src 'self' 'unsafe-inline'",
      // Styles: same origin + inline styles (used by Tailwind/emotion)
      "style-src 'self' 'unsafe-inline' https://fonts.googleapis.com",
      // Fonts from Google Fonts CDN
      "font-src 'self' https://fonts.gstatic.com",
      // Images: same origin + Substack CDN + data URIs (for inline SVGs)
      "img-src 'self' data: https://*.substack.com https://substackcdn.com https://*.substackcdn.com",
      // API / fetch targets: same origin + Supabase
      "connect-src 'self' https://*.supabase.co wss://*.supabase.co",
      // No objects, embeds, or applets
      "object-src 'none'",
      // Base tag restricted to same origin
      "base-uri 'self'",
      // All form submissions must target same origin
      "form-action 'self'",
    ].join("; ");

    res.setHeader("Content-Security-Policy", csp);

    // Prevents this window from gaining access to cross-origin opener windows (e.g. popup attacks).
    res.setHeader("Cross-Origin-Opener-Policy", "same-origin");

    // Enforce HTTPS in production — browsers will refuse to connect over HTTP for 1 year.
    if (process.env.NODE_ENV === "production") {
      res.setHeader(
        "Strict-Transport-Security",
        "max-age=31536000; includeSubDomains; preload"
      );
    }

    next();
  };
}
